src/System.Management.Automation/engine/parser/Compiler.cs

```diff
@@ -1931,6 +1931,14 @@ private enum CaptureAstContext
             Ast ast,
             CaptureAstContext context,
             MergeRedirectExprs generateRedirectExprs = null)
```

Comment:
This change is no longer necessary, right?

Reply:
It's still required. `Compile(ast)` for an expression ast doesn't write the expression value to pipe, so I need to do it separately, as in `CaptureAstResults(CallAddPipe(Compile(ternaryExpressionAst.Condition), s_getCurrentPipe), context, ...)`

Reply:
Is there a reason the condition is not a `PipelineBaseAst` like conditions in an `if`/`while`/`for`/`switch`?

Reply:
Yes, the reason is both `?` and `:` are valid arguments for a function or cmdlet, e.g. `get-command ?`.
So, it's impossible to allow the ternary condition to accept a `PipelineBaseAst` directly because the `?` char needs to be parse as a argument to the function/cmdlet just like how it works today.
Therefore, I make the ternary condition an `ExpressionAst`, so you will need to use the paren expression if you want to use a cmdlet as the condition, `(Test-Path $mypath) ? ... : ...`

Reply:
What about something like `(Test-Path $p) ? ... : ...`?

Reply:
Then the second `?` will be considered as another ternary expression, because we will allow
```
<condition-1>
       ? <condition-2> ? ... : ...
       : <condition-3> ? ... : ...
```
Example:
```
PS:2> $false ? 2 : $true ? 3 : 4
3
```

Reply:
The second `?` was punctuation. I was just pointing out a useful example where the condition is not an expression, and I don't think you should be forced to write it this way: `$(Test-Path $p) ? ... : ...`

Reply:
> The second ? was punctuation.

Oh, sorry that I misread it.
I don't force user to use the sub expression. Parenthesis expression works fine in this case:
```
PS:7> (Test-Path $pwd) ? "exist" : "non-exist"
exist
```